Move processing of messages in status-go & introduce protobuf

cammellos opened the pull request:
Move message processing to status-go and introduce protobuf
    
This commit moves all the processing of messages to status-go.
    
Messages are going arrive to status-react already saved and processed.

Receiving/sending/retrieving from db is now using the same identical structure. The only processing left in status-react is to mark the messages as seen and update the unviewed count locally (only status-react knows whether the count should be updated).
    
Partially remove commands as well as won't be used anymore.

Requires manual QA (sending messages in all types of chats, blocking users, i.e anything related with chat).

Some tests are failing as now system-messages in group chat display markdown correctly, I have updated the tests and should pass if run against this branch.

I will replace status-go version with a proper one once the corresponding PR is merged  https://github.com/status-im/status-go/pull/1706

status: ready

cammellos committed:
Move message processing to status-go and introduce protobuf

This commit moves all the processing of messages to status-go.

Messages are going arrive to status-react already saved an processed.

Receiving/sending/retrieving from db is now using the same identical
structure. The only processing left in status-react is to mark the
messages as seen and update the unviewed count locally (only
status-react knows whether the count should be updated).

Partially remove commands as well as won't be used anymore.

Signed-off-by: Andrea Maria Piana <andrea.maria.piana@gmail.com>

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :x: | 9ab32783 | [#1](https://ci.status.im/job/status-react/job/prs/job/ios/job/PR-9547/1/) | 2019-11-28 16:12:11 | ~48 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-react/job/prs/job/ios/job/PR-9547/1/consoleText) |

cammellos commented on src/status_im/contact/core.cljs:
```diff
@@ -109,6 +109,7 @@
   ;; when using multi-device & clocks are out of sync
   ;; Using logical clocks is probably the correct way to handle it, but an overkill
   ;; for now
+  (println "HANDLING CONTACT UPDATE" timestamp public-key name profile-image address)
```
Remove println

yenda commented on test/cljs/status_im/test/data_store/messages.cljs:
```diff
@@ -6,60 +6,43 @@
-                     :outgoing :sending
-                     :message-type :public-group-user-message
+                     :outgoing true
+                     :message-type 0
```
maybe we could turn this into a keyword during the unmarshalling instead of using a constant?

cammellos replied:
I kind of like using constants, it's a bit of nightmare to go around changing up things when the value is used directly (and prone to errors, I think we have a wrong keyword somewhere)

yenda commented on src/status_im/transport/message/core.cljs:
```diff
@@ -91,18 +52,48 @@
-        ;; it will keep processing events ignoring user input.
-        :dispatch-later [{:ms 20 :dispatch [::process remaining-messages now-in-s]}])
-       (receive-message cofx now-in-s chat-id message)))))
+ ::process-2
```
why is it process-2?

cammellos replied:
I had `:process` before, so didn't override it, but removed it after, and forgot to rename :)

yenda commented on env/dev/env/config.cljs:
```diff
@@ -1,6 +1,6 @@
 (ns env.config)
```
remove this file

cammellos commented:
@churik @Serhy I have updated a test as group chats are not synced anymore when sending a message.

That's because we embed a message in the membership updates, and I moved messages to protobuf, so I need to move membership updates to protobuf to fix it, but doing so in a single PR would have been too much work, and keeping compatibility would have also being too much work before the code freeze. I will work on that this week and uncomment the test once that's fixed, hope is ok.

The rest of the failures don't see to be related to this PR. When you have time could you have a look at the last build just to smoke check everything works ok?
Thanks!

churik commented:
@cammellos 

### Issue 1: new messages are received from blocked users
**Steps:**
- as user A and B join to public chat
- as User A: send several messages to public chats
- as User B: open public chat and block User A
- as User A: send several new messages
**Expected result:**
new messages are not received
**Actual result:**
new messages are received

cammellos commented:
@churik thanks, it should be fixed in the last build.

churik commented:
### Issue 2: No preview for chats in home after relogin (reproducibility ~20%)

After that issue preview for all chats are not loaded until new messages arrived.
Couldn't catch on IOS, but I'm not sure that it is platform-specific.

![screencast 2019-12-03 13-23-54](https://user-images.githubusercontent.com/4557972/70050937-5dd76700-15d0-11ea-9195-63f2f3aabdf7.gif)


**Steps:**
- create multi account and join to several chats with history
- several relogin attempts (profile -> logout -> login)

**Expected result:**
preview are preserved

**Actual result:**
"no messages" in all chats

ADB Log:  [no-messages.log](https://github.com/status-im/status-react/files/3916514/no-messages.log)
All logs: [Status-debug-logs.zip](https://github.com/status-im/status-react/files/3916523/Status-debug-logs.zip)

Serhy commented:
### Issue 3: 
Messages sent in group chat after ContactA invitation are visible to ContactA on croup chat invitation screen:
1) Open Status and create two accounts UserA and UserB
2) UserA adds UserB in contacts
3) UserA creates group chat with UserB (Note: at this point UserB receives invitation to Group Chat on Chats screen)
4) UserA sends text messages in Group chat
**Actual result:** All messages are already visible in preview of thr last message Group chat item of Chats screen 
5) UserB opens Group chat `UserA invited you to... ` screen
**Actual result:** 
<img width="343" alt="Screenshot 2019-12-03 at 15 26 57" src="https://user-images.githubusercontent.com/8749671/70055482-4903cf00-15e2-11ea-992a-9f7df5303f5e.png">



### Issue 4:
I can send messages in Group chat with only 1 participant and they have endless spinner. In develop build  it's also possible to send a message in group chat with only 1 member but nothing is sent in fact.

1) Create an account (UserA)
2) Add UserB in contacts
3) Create Group chat with UserB
4) UserB leaves group chat (decline invitation)
5) UserA send a text messages in Group chat
**Expected result:** Tapping 'Send' button clears message input
**Actual result:** Message is sent with endless spinner next to message

<img width="353" alt="Screenshot 2019-12-03 at 15 42 20" src="https://user-images.githubusercontent.com/8749671/70056202-b82df300-15e3-11ea-898f-7b6b30ba1696.png">

Serhy commented:
@cammellos you seem not mentioned above, but is it expected this changes are kind of 'braking' changes between develop? Messaging doesn't work between nightly develop builds and PR builds.

Taking into account V1 itself will be a "breaking change" for Beta version, but just wondering whether it's expected here between develop and PR?

cammellos commented:
@Serhy yes, expected, this is a breaking change between develop & this PR, thanks for the feedback btw

churik commented:
### Issue 4: Messages from offline are not fetched for group chat (reproducibility ~50%)

Messages have never arrived sometimes, and "Fetch last 24 hours" doesn't help.
![no-messages-fetching](https://user-images.githubusercontent.com/4557972/70057493-81ed7500-15dd-11ea-8b7d-df9b35dfbcb1.png)
So these messages are lost.

**Steps:**
- as User A: create group chat with User B
- as User B: join group chat and public chat and exchange several messages
- as User B: close app
- as User A: send several messages to both public and group chats
- as User B: open app and login, wait for messages

**Expected result:**
all messages have arrived

**Actual result:**
messages for public chat are arrived, for group - no.
ADB Log: [no-messages-fetched.log](https://github.com/status-im/status-react/files/3916983/no-messages-fetched.log)

Geth log: [geth.log](https://github.com/status-im/status-react/files/3916986/geth.log)

churik commented:
### Issue 6: Sorting is not applied for the message sender


**Steps:**
-  create group chat, public chat, 1-1 chat
- send message 1 to  group chat, 2 to  public chat, 3 to 1-1 chat
- go back to home view

**Expected result:**
chats are sorted by last sent message

**Actual result:**
![sorting](https://user-images.githubusercontent.com/4557972/70058851-fa553580-15df-11ea-9216-fcbd25787ee1.gif)

cammellos commented:
@churik @Serhy :
1,2,3) Should be fixed
4) I left it as it is for now, we probably should decided what to do (seems reasonable not to allow sending messages if no users have joined a group chat), but otherwise if it's fine with you I will address separately
5) Is also in develop and as discussed will be tracked separately
6) Should be fixed

Thanks!

churik commented:
@cammellos 
one thing that I see from full set of e2e tests is random crash on tapping "New chat" window.
Full video / logcat is [here](https://app.saucelabs.com/tests/43527423de714648881de6508cbf82b4?auth=e3f92a1ede922c99b2b49273fe05fa0d)

Crash logs:
```
12-04 07:05:33.317  4351  4377 D ReactNativeJS: DEBUG [status-im.ui.screens.routing.core:34] - :on-will-focus :new-chat
12-04 07:05:33.323  4351  4351 E JavaBinder: !!! FAILED BINDER TRANSACTION !!!  (parcel size = 448)
12-04 07:05:33.323  4351  4351 D AndroidRuntime: Shutting down VM
12-04 07:05:33.324  1636  1704 W InputDispatcher: channel '264c501 im.status.ethereum.pr/im.status.ethereum.MainActivity (server)' ~ Consumer closed input channel or an error occurred.  events=0x9
12-04 07:05:33.324  1636  1704 E InputDispatcher: channel '264c501 im.status.ethereum.pr/im.status.ethereum.MainActivity (server)' ~ Channel is unrecoverably broken and will be disposed!
--------- beginning of crash
12-04 07:05:33.324  4351  4351 E AndroidRuntime: FATAL EXCEPTION: main
12-04 07:05:33.324  4351  4351 E AndroidRuntime: Process: im.status.ethereum.pr, PID: 4351
12-04 07:05:33.324  4351  4351 E AndroidRuntime: java.lang.RuntimeException: Adding window failed
12-04 07:05:33.324  4351  4351 E AndroidRuntime:  at android.view.ViewRootImpl.setView(ViewRootImpl.java:738)
12-04 07:05:33.324  4351  4351 E AndroidRuntime:  at android.view.WindowManagerGlobal.addView(WindowManagerGlobal.java:356)
12-04 07:05:33.324  4351  4351 E AndroidRuntime:  at android.view.WindowManagerImpl.addView(WindowManagerImpl.java:92)
12-04 07:05:33.324  4351  4351 E AndroidRuntime:  at android.app.Dialog.show(Dialog.java:330)
```
I didn't see this earlier and suppose it can be related most likely to https://github.com/status-im/status-react/pull/9571, but not to your changes.
Can you confirm it?

churik commented:
### Issue 7: On fetching previous history got error Unsupported top level event type (IOS)

Also can be related to RN upgrade.

**Steps:**
-  join public chat #kuku 
- fetch previous history for 02/12

**Expected result:**
no errors

**Actual result:**
![IMG_D5EDB3F7DCE4-1](https://user-images.githubusercontent.com/4557972/70162826-a1a89a00-16be-11ea-8750-d30fb3c2a671.jpeg)

**Logs:** 
[status_logs-1.zip](https://github.com/status-im/status-react/files/3922986/status_logs-1.zip)

cammellos commented:
@churik both seems to be related to react-native upgrade I believe, as this parts of the codebase were not touched, I can try to replicate the start-chat issue, (I don't have an ios device)

churik commented:
@cammellos 
got it. so will log separately both of them.

churik commented:
@cammellos 
### Issue 8: PNs are not working (PR <-> PR)

PNs are working in nightly <-> nightly and not working PR <-> PR

**Steps:**
-  device A: start 1-1 chat with device B (Android)
- device B: turn on Notifications in profile. "Backgroung notification service" PN is appeared.
- device A: send several messages to  1-1 chat with device B

**Expected result:** PNs on device B

**Actual result:** no PNs

**Logs:**  [no-pns.log](https://github.com/status-im/status-react/files/3926465/no-pns.log)

churik commented:
### Issue 9: 
Error at attempt to add members to group chat when you have been made admin 

The issue is not reproducible on nightly.

1) Open Status and create two accounts UserA and UserB
2) UserA adds UserB in contacts
3) UserA creates group chat with UserB 
4) UserA made  UserB admin
5) UserB open group chat > group info > Add members 

**Expected result:**  can add users
**Actual result:** 
![photo_2019-12-05 12 19 35](https://user-images.githubusercontent.com/4557972/70230857-8474d980-1759-11ea-9da2-1276c30ac4c4.jpeg)
**Logs:**
[Status-debug-logs.zip](https://github.com/status-im/status-react/files/3926543/Status-debug-logs.zip)

churik commented:
All mentioned issues are fixed.
All test cases and results are here: 
[testrail-report-19.pdf](https://github.com/status-im/status-react/files/3927802/testrail-report-19.pdf)

@cammellos thank you for great work!
